ra/registration-authority.go

```diff
@@ -790,38 +790,34 @@ func (ra *RegistrationAuthorityImpl) UpdateAuthorization(base core.Authorization
 
 	// Dispatch to the VA for service
 
-	doValidate := func() {
-		records, err := ra.VA.PerformValidation(authz.Identifier.Value, authz.Challenges[challengeIndex], authz)
-		var prob *probs.ProblemDetails
-		if p, ok := err.(*probs.ProblemDetails); ok {
-			prob = p
-		} else if err != nil {
-			// TODO something more sensible than this, please
-			return
-		}
-
-		// Save the updated records
-		chal := &authz.Challenges[challengeIndex]
-		chal.ValidationRecord = records
-		if prob != nil {
-			chal.Status = core.StatusInvalid
-			chal.Error = prob
-		} else if !chal.RecordsSane() {
-			chal.Status = core.StatusInvalid
-			chal.Error = &probs.ProblemDetails{Type: probs.ServerInternalProblem,
-				Detail: "Records for validation failed sanity check"}
-		} else {
-			chal.Status = core.StatusValid
-		}
-		authz.Challenges[challengeIndex] = *chal
+	if !ra.useNewVARPC {
+		ra.VA.UpdateValidations(authz, challengeIndex)
+	} else {
+		go func() {
+			records, err := ra.VA.PerformValidation(authz.Identifier.Value, authz.Challenges[challengeIndex], authz)
+			var prob *probs.ProblemDetails
+			if p, ok := err.(*probs.ProblemDetails); ok {
+				prob = p
+			} else if err != nil {
+				prob = probs.ServerInternal("uncaught error in RPC to VA")
+			}
 
-		ra.OnValidationUpdate(authz)
-	}
+			// Save the updated records
+			chal := &authz.Challenges[challengeIndex]
+			chal.ValidationRecord = records
+			if prob != nil {
+				chal.Status = core.StatusInvalid
+				chal.Error = prob
+			} else if !chal.RecordsSane() {
+				chal.Status = core.StatusInvalid
+				chal.Error = probs.ServerInternal("Records for validation failed sanity check")
+			} else {
+				chal.Status = core.StatusValid
+			}
+			authz.Challenges[challengeIndex] = *chal
 
-	if ra.useNewVARPC {
-		go doValidate()
-	} else {
-		ra.VA.UpdateValidations(authz, challengeIndex)
+			ra.OnValidationUpdate(authz)
+		}()
 	}
 
 	ra.stats.Inc("RA.UpdatedPendingAuthorizations", 1, 1.0)
```

Comment:
This should be moved into the goroutine above since it will now be invoked before the validation actually takes place.